### DISCUSSION TITLE
iOS Native App Integration

### DISCUSSION THREAD

**Discussion by akozlik** [MasterKale/SimpleWebAuthn#411] [General]
Has anyone been able to get this to work with a native iOS app? The challenge in the clientDataJSON property that's sent back after creating a credential is base64 encoded by Apple but this library's validation seems to expect it to be decoded. 

Are there any good docs for integrating with this natively? All the resources I've found are all over the place and nothing seems consistent.

**Comment by MasterKale** [comment]
Hmm, I've not come across this issue for all the iPhone registration I see at my day job. Perhaps the challenge is being double-encoded somewhere so what seems like the challenge in the response being sent back is base64-encoded is actually an erroneous additional base64 encoding of the challenge prior to handing it to the authenticator. 🤔

**Reply by akozlik** [reply]
It's definitely odd. I'll post some code samples a bit later this morning. This library has been incredibly helpful but this one thing has been frustrating. Apple's docs are supremely not helpful either haha.

**Reply by akozlik** [reply]
Hey Matt,

I dug into this with my partner today and I think I have more info if you're able to help. For context we're currently dealing with user registration.

**Server Generates Challenge**
Client POSTs to /users endpoint to create a new user. I create a record for the user in our DB and get an associated user ID. I use that user ID, generate a challenge with the simplewebauthn library, and return the raw response back to the client.

```
{
    challenge: 'FwNK50NbyMmTDtcnseKSNrXWDu356Fz04ByfXHxl3C0',
    rp: { name: 'myname', id: 'mydomain' },
    user: {
    id: 'dbID',
    name: 'theemail',
    displayName: 'thedisplayname'
    },
    pubKeyCredParams: [
    { alg: -8, type: 'public-key' },
    { alg: -7, type: 'public-key' },
    { alg: -257, type: 'public-key' }
    ],
    timeout: 60000,
    attestation: 'direct',
    excludeCredentials: [],
    authenticatorSelection: {
    residentKey: 'preferred',
    userVerification: 'preferred',
    requireResidentKey: false
    },
    extensions: { credProps: true }
}
```

I store the challenge `FwNK50NbyMmTDtcnseKSNrXWDu356Fz04ByfXHxl3C0 ` server-side as a string object associated with the user ID `dbID`. This is stored as a plain string and is not encoded.

**iOS Generates Credential**
The iOS client receives this data and passes the challenge to Apple to generate a credential.

```
let registrationRequest = publicKeyProvider.createCredentialRegistrationRequest(
            challenge: Data(challenge.utf8),
            name: username,
            userID: Data(userID.utf8)
        )
```

Please note that Apple's `createCredentialRegistrationRequest` method requires the challenge be sent in as a `Data` object. This explains why it's being returned to me in `clientDataJSON` as a data object since that's what iOS is receiving.

After registering the credential my partner sends me the data to finalize the registration. He is not manipulating any of this data and is sending it to me just as iOS provides.

**Server Validates Credential**
The decoded clientDataJSON I receive looks like:

```
{
    "type": "webauthn.create",
    "challenge": "RndOSzUwTmJ5TW1URHRjbnNlS1NOclhXRHUzNTZGejA0QnlmWEh4bDNDMA",
    "origin": "myOrigin"
}
```

Note that the `challenge` property is the `Data` representation of the original challenge I sent back which was `FwNK50NbyMmTDtcnseKSNrXWDu356Fz04ByfXHxl3C0`. This can be confirmed by decoding the string.

When the client sends this data up I'm comparing the Apple encoded Data object to the expectedChallenge plain string representation I've stored. Obviously there's a mismatch so the validation fails.

I tried storing the original challenge using the `btoa()` JS function but that function adds padding which results in the challenge being stored as `RndOSzUwTmJ5TW1URHRjbnNlS1NOclhXRHUzNTZGejA0QnlmWEh4bDNDMA==`

It's close but still a mismatch so the validation fails. I want to avoid hardcoding the removal of the '==' at the end of the data since there was no indication in the documentation that this was necessary.

**Questions**
- How should I store the original challenge for validation? Do I need to encode it somehow? 
  - Your documentation had a pseudocode example calling a `setUserCurrentChallenge` function but there were no specifics about how the data should be stored. Your sample used string which is what we originally stored as.
- Are you aware of anything anybody has had to do to the string before sending it into Apple's method as a `Data` object?
- Is there anything obviously wrong that I'm doing?
- If I do have to encode the challenge before storing it what's the best way to do so?

Very sorry for the long post but I'm at my wit's end trying to figure this out. I'm sure there's something that we're doing wrong but I'm having trouble figuring out what that is. Any additional guidance would be greatly appreciated. Happy to pay a consulting fee to hop on a call as well if that's something you'd prefer.

Thank you in advance for any help you can give.

**Reply by akozlik** [reply]
I thought it might be worth noting when I say "Native iOS" I mean the passkey libraries in a native app. It's not a reference to Safari on iOS.

**Reply by MasterKale** [reply]
I think I see what's going on here: you're passing the UTF-8 bytes of the **already encoded challenge** to the authenticator which then encodes those UTF-8 bytes into a base64url string as per the spec when it collects `clientDataJSON`:

https://www.w3.org/TR/webauthn-2/#dom-collectedclientdata-challenge

If you look at the corresponding `startRegistration()` in **@simplewebauthn/browser** you'll see that this method takes care of this before passing it into the WebAuthn API call:

https://github.com/MasterKale/SimpleWebAuthn/blob/master/packages/browser/src/methods/startRegistration.ts#L32

I think you can fix this by changing `Data(challenge.utf8)` to instead, on the iOS side of things, decode the base64url-encoded string `challenge` back into the original bytes before passing it into `publicKeyProvider.createCredentialRegistrationRequest()`. Then the `challenge` inside `clientDataJSON` should start matching the string coming out of `generateRegistrationOptions()`.

**Reply by akozlik** [reply]
Thanks for the reply! This was great guidance.

We took a look and found that your startRegistration method is calling a base64 encoding helper method that's adding padding. Turns out that Apple doesn't do this so we had to add the padding on our end. After padding the data we had great success and the challenges matched! Funny thing is that we thought this was hacking something together on our end but was actually needed.

I think the key thing here is to note for others who find this thread is that you need to pad the client JSON if you're using native iOS.

The challenge is now validating correctly, but it looks like the hashed RP ID is not matching in the attestation object. We're trying to debug this but running up against a wall. 

On iOS we create a credential provider and pass in the same RP ID that's returned in the challenge data.

```
let publicKeyProvider = ASAuthorizationPlatformPublicKeyCredentialProvider(
            relyingPartyIdentifier: decodedResponse.relayingParty.id
        )
```

This is just a string representation of our domain. We've logged this and confirmed it's correct.

When we run `verifyRegistrationResponse` we pass in an `expectedRPID: "stageapi.domain.com"` which we've confirmed as the same value as what's being registered with Apple.

The verification process fails due to an "Unexpected RP ID hash"

```
2023-07-13T20:00:10.421187+00:00 app[web.1]: UnexpectedRPIDHash: Unexpected RP ID hash
2023-07-13T20:00:10.421188+00:00 app[web.1]: at matchExpectedRPID (/app/.next/server/chunks/479.js:6403:19)
2023-07-13T20:00:10.421189+00:00 app[web.1]: at async verifyRegistrationResponse (/app/.next/server/chunks/479.js:8500:9)
2023-07-13T20:00:10.421189+00:00 app[web.1]: at async POST (/app/.next/server/app/v1/users/challenge/finish/route.js:239:24)
2023-07-13T20:00:10.421189+00:00 app[web.1]: at async /app/.next/server/chunks/981.js:2823:37
```

We're simply passing in the plain text string "stageapi.domain.com" to APPLE. Nothing is being encoded or modified on our end. I verified this is the same RP ID that's being passed by the browser library and the browser does not have any issues.

We have attestation type set to "none" when we create the challenge through our API.

```
const options = generateRegistrationOptions({
        rpName,
        rpID,
        userID: user._id.toString(),
        userName: user.username,
        attestationType: 'none',
      });
```

On the iOS side we also have attestation set to none when we create the object to generate a cert.

I'm not sure if I'm misunderstanding but if we have attestationType set to none shouldn't the attestation check be skipped?

If you have any suggestions or can push us in the right direction I'd really appreciate it. Happy to provide other code samples if there's something more helpful to see.

**Reply by akozlik** [reply]
Some more fire for the mystery.

We did a CBOR decode on the attestationObject that was sent from the browser as well as sent from native iOS.

**Web authData in attestationObject**
`authData: <Buffer 27 ce ba fa 48 d4 c6 14 5a 33 3b e1 93 8e 3f 4a ca 8e 46 de 4d 11 f2 70 f7 0a 50 ae 3d 39 87 e9 5d 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 102 more bytes>`

**iOS authData in attestationObject**
`authData: <Buffer 27 ce ba fa 48 d4 c6 14 5a 33 3b e1 93 8e 3f 4a ca 8e 46 de 4d 11 f2 70 f7 0a 50 ae 3d 39 87 e9 5d 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 102 more bytes>`

Web authData works fine but iOS authData fails the check. Those first 32 bytes are supposed to be the hashed RP ID and you can see they're the same. Instead the iOS one throws the RP ID mismatch error.

The code I used to print this out was:
```
let attestationObjectBuffer = base64url.toBuffer(attestationObject);
let ctapMakeCredResp        = cbor.decodeAllSync(attestationObjectBuffer)[0];
console.log(ctapMakeCredResp);
```

I hope this helps and thanks again for your help on this so far. Totally understand if you need to bail on the issue but just hitting a wall on our side.

**Reply by MasterKale** [reply]
The hashes definitely don't match lol

```js
Buffer.from([
  0x27, 0xce, 0xba, 0xfa, 0x48, 0xd4, 0xc6,
  0x14, 0x5a, 0x33, 0x3b, 0xe1, 0x93, 0x8e,
  0x3f, 0x4a, 0xca, 0x8e, 0x46, 0xde, 0x4d,
  0x11, 0xf2, 0x70, 0xf7, 0x0a, 0x50, 0xae,
  0x3d, 0x39, 0x87, 0xe9,
]).toString('base64');
// J866+kjUxhRaMzvhk44/SsqORt5NEfJw9wpQrj05h+k=

Buffer.from('stageapi.domain.com', 'utf-8').toString('base64');
// c3RhZ2VhcGkuZG9tYWluLmNvbQ==
```

🤔

**Reply by akozlik** [reply]
Whoops sorry if that wasn't clear. The buffer I posted is my actual domain's hash. The domain.com string was just a sample domain for explanation's purpose. I'll double check against the real domain but they should match.

**Reply by akozlik** [reply]
It'll be easier if I just put in my domain and I'll remove it later. 

The RP ID we're using is `stageapi.getsecretstash.com`.

When I hash that string with sha256 I get `27cebafa48d4c6145a333be1938e3f4aca8e46de4d11f270f70a50ae3d3987e9`. Breaking that into bytes yields

```
[
  0x27, 0xce, 0xba, 0xfa, 0x48, 0xd4, 0xc6,
  0x14, 0x5a, 0x33, 0x3b, 0xe1, 0x93, 0x8e,
  0x3f, 0x4a, 0xca, 0x8e, 0x46, 0xde, 0x4d,
  0x11, 0xf2, 0x70, 0xf7, 0x0a, 0x50, 0xae,
  0x3d, 0x39, 0x87, 0xe9,
]
```

This matches what I previously posted from the authData and matches your byte sample.

I ran

```
let data = Buffer.from('stageapi.getsecretstash.com', 'utf-8').toString('base64');
console.log(data);
```

And this yielded `c3RhZ2VhcGkuZ2V0c2VjcmV0c3Rhc2guY29t` which doesn't match. However, I believe in your example we needed to SHA 256 the string first before comparing to the authData bytes since authData houses the hashed value.

This led me to 

```
function hash(string) {
    return createHash('sha256').update(string).digest('hex');
}

console.log("Hashed: " + hash('stageapi.getsecretstash.com'));
```

The result was the same hashed string from before 27cebafa48d4c6145a333be1938e3f4aca8e46de4d11f270f70a50ae3d3987e9. Passing that into your Buffer example:

```
console.log(Buffer.from(hash(rpid), 'hex').toString('base64'))
```

Yields `J866+kjUxhRaMzvhk44/SsqORt5NEfJw9wpQrj05h+k=` which matches the original authData.

So it looks like the correct RP ID is being passed through on both examples but for some reason they're being thrown as a mismatch.

I just ran another test from an iOS device and generated the following attestation value. I've also included the broken down data to show that the data is correct.

**Failed Native iOS Validation**
```
Attestation Object: o2NmbXRkbm9uZWdhdHRTdG10oGhhdXRoRGF0YViYJ866+kjUxhRaMzvhk44/SsqORt5NEfJw9wpQrj05h+ldAAAAAAAAAAAAAAAAAAAAAAAAAAAAFDs0aGW2IFefWp2YBTnMF3frtuDqpQECAyYgASFYIEdsXfmoFVtAxie5WbvcSZBbL/XIvpzRFx7VdoY3S1Q4IlggB/+Nb7vAg4sKmDF79GMTdTmxn0CFhZxpKV74WuTuFbc=

Attestation Object Buffer
<Buffer a3 63 66 6d 74 64 6e 6f 6e 65 67 61 74 74 53 74 6d 74 a0 68 61 75 74 68 44 61 74 61 58 98 27 ce ba fa 48 d4 c6 14 5a 33 3b e1 93 8e 3f 4a ca 8e 46 de ... 132 more bytes>

ctap
{
  fmt: 'none',
  attStmt: {},
  authData: <Buffer 27 ce ba fa 48 d4 c6 14 5a 33 3b e1 93 8e 3f 4a ca 8e 46 de 4d 11 f2 70 f7 0a 50 ae 3d 39 87 e9 5d 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 102 more bytes>
}

auth data
{
  rpIdHash: <Buffer 27 ce ba fa 48 d4 c6 14 5a 33 3b e1 93 8e 3f 4a ca 8e 46 de 4d 11 f2 70 f7 0a 50 ae 3d 39 87 e9>,
  counter: 0,
  flags: { up: true, uv: true, at: true, ed: false, flagsInt: 93 },
  counterBuffer: <Buffer 00 00 00 00>,
  aaguid: '00000000000000000000000000000000',
  credIdBuffer: <Buffer 3b 34 68 65 b6 20 57 9f 5a 9d 98 05 39 cc 17 77 eb b6 e0 ea>,
  cosePublicKeyBuffer: <Buffer a5 01 02 03 26 20 01 21 58 20 47 6c 5d f9 a8 15 5b 40 c6 27 b9 59 bb dc 49 90 5b 2f f5 c8 be 9c d1 17 1e d5 76 86 37 4b 54 38 22 58 20 07 ff 8d 6f bb ... 27 more bytes>,
  coseExtensionsDataBuffer: undefined
}
```

Again, you can see the rpidHash matches what's above. Odd right?

Just for S&Gs I ran the same test with the attestation object I get back from the browser which verifies correctly. All he data is the same except for the `cosePublicKeyBuffer`. Maybe we need to specify to use a certain public key for native iOS authentication? I'm not sure how we would do that or if I'm just speaking gibberish.

**Successful Web Browser Validation**
```
Attestation Object: o2NmbXRkbm9uZWdhdHRTdG10oGhhdXRoRGF0YViYJ866-kjUxhRaMzvhk44_SsqORt5NEfJw9wpQrj05h-ldAAAAAAAAAAAAAAAAAAAAAAAAAAAAFASdzaxl7nKt3X6PYMT8kq9MpQuWpQECAyYgASFYIJXzINGJElMN3Qa2p2VXkjnopWXT9c8Rrv4sUFnuz_PRIlgghQ2IpVRSGw-DcjdHD7-_0AgEGgVpgfVzP8PBiVhBGII

Buffer
<Buffer a3 63 66 6d 74 64 6e 6f 6e 65 67 61 74 74 53 74 6d 74 a0 68 61 75 74 68 44 61 74 61 58 98 27 ce ba fa 48 d4 c6 14 5a 33 3b e1 93 8e 3f 4a ca 8e 46 de ... 132 more bytes>

ctap
{
  fmt: 'none',
  attStmt: {},
  authData: <Buffer 27 ce ba fa 48 d4 c6 14 5a 33 3b e1 93 8e 3f 4a ca 8e 46 de 4d 11 f2 70 f7 0a 50 ae 3d 39 87 e9 5d 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 ... 102 more bytes>
}

auth data
{
  rpIdHash: <Buffer 27 ce ba fa 48 d4 c6 14 5a 33 3b e1 93 8e 3f 4a ca 8e 46 de 4d 11 f2 70 f7 0a 50 ae 3d 39 87 e9>,
  counter: 0,
  flags: { up: true, uv: true, at: true, ed: false, flagsInt: 93 },
  counterBuffer: <Buffer 00 00 00 00>,
  aaguid: '00000000000000000000000000000000',
  credIdBuffer: <Buffer 04 9d cd ac 65 ee 72 ad dd 7e 8f 60 c4 fc 92 af 4c a5 0b 96>,
  cosePublicKeyBuffer: <Buffer a5 01 02 03 26 20 01 21 58 20 95 f3 20 d1 89 12 53 0d dd 06 b6 a7 65 57 92 39 e8 a5 65 d3 f5 cf 11 ae fe 2c 50 59 ee cf f3 d1 22 58 20 85 0d 88 a5 54 ... 27 more bytes>,
  coseExtensionsDataBuffer: undefined
}
```

Any suggestions? I'm looking at potentially forking this and pulling it into my own repo so I can add some logging on my server and start trying to look at what data is coming into this library. This is definitely strange.

**Reply by MasterKale** [reply]
> And this yielded `c3RhZ2VhcGkuZ2V0c2VjcmV0c3Rhc2guY29t` which doesn't match. However, I believe in your example we needed to SHA 256 the string first before comparing to the authData bytes since authData houses the hashed value.

I can't believe I forgot to hash the RP ID, it's been a long week 🤦

That aside I think I figured it out: the **native** `attestationObject` is base64-encoded, while SimpleWebAuthn expects base64url when parsing the attestation object. When I decode the attestation object as base64 then the native app response validates the RP ID just fine:

```ts
(async () => {
  // Native
  const attestationObject = 'o2NmbXRkbm9uZWdhdHRTdG10oGhhdXRoRGF0YViYJ866+kjUxhRaMzvhk44/SsqORt5NEfJw9wpQrj05h+ldAAAAAAAAAAAAAAAAAAAAAAAAAAAAFDs0aGW2IFefWp2YBTnMF3frtuDqpQECAyYgASFYIEdsXfmoFVtAxie5WbvcSZBbL/XIvpzRFx7VdoY3S1Q4IlggB/+Nb7vAg4sKmDF79GMTdTmxn0CFhZxpKV74WuTuFbc=';

  console.log('\nParsing native attestation object as base64url')
  const decoded1 = decodeAttestationObject(isoBase64URL.toBuffer(attestationObject));
  const parsedAuthData1 = parseAuthenticatorData(decoded1.get('authData'));
  console.log('response hex RP ID hash:', isoUint8Array.toHex(parsedAuthData1.rpIdHash));
  console.log(
    'computed hex RP ID hash:',
    isoUint8Array.toHex(
      await isoCrypto.digest(isoUint8Array.fromUTF8String('stageapi.getsecretstash.com'), COSEALG.ES256),
    ),
  );

  try {
    await matchExpectedRPID(parsedAuthData1.rpIdHash, ['stageapi.getsecretstash.com']);
    console.log('found match');
  } catch (err) {
    console.error(err);
  }

  console.log('\nParsing native attestation object as base64')
  const decoded2 = decodeAttestationObject(isoBase64URL.toBuffer(attestationObject, 'base64'));
  const parsedAuthData2 = parseAuthenticatorData(decoded2.get('authData'));
  console.log('response hex RP ID hash:', isoUint8Array.toHex(parsedAuthData2.rpIdHash));
  console.log(
    'computed hex RP ID hash:',
    isoUint8Array.toHex(
      await isoCrypto.digest(isoUint8Array.fromUTF8String('stageapi.getsecretstash.com'), COSEALG.ES256),
    ),
  );

  try {
    await matchExpectedRPID(parsedAuthData2.rpIdHash, ['stageapi.getsecretstash.com']);
    console.log('found match');
  } catch (err) {
    console.error(err);
  }
})();
```

The output:

```
Parsing native attestation object as base64url
response hex RP ID hash: 27ceba0248d4c6145a333be1938e004aca8e46de4d11f270f70a50ae3d398409
computed hex RP ID hash: 27cebafa48d4c6145a333be1938e3f4aca8e46de4d11f270f70a50ae3d3987e9
UnexpectedRPIDHash: Unexpected RP ID hash
    at matchExpectedRPID (/Users/matt/Developer/simplewebauthn/packages/server/src/helpers/matchExpectedRPID.ts:30:13)
    at async /Users/matt/Developer/simplewebauthn/packages/server/attestationobject.ts:32:5

Parsing native attestation object as base64
response hex RP ID hash: 27cebafa48d4c6145a333be1938e3f4aca8e46de4d11f270f70a50ae3d3987e9
computed hex RP ID hash: 27cebafa48d4c6145a333be1938e3f4aca8e46de4d11f270f70a50ae3d3987e9
found match
```

Give me a sec to think of a path forward for you...

**Reply by MasterKale** [reply]
One thing you can do is use `isoBase64URL.isBase64url(attestationObject)` to determine if `attestationObject` is a base64url string or not. If it _isn't_, then convert it to base64url before feeding the response into `verifyRegistrationResponse()`:

```ts
(async () => {
  // Native
  let attestationObject = 'o2NmbXRkbm9uZWdhdHRTdG10oGhhdXRoRGF0YViYJ866+kjUxhRaMzvhk44/SsqORt5NEfJw9wpQrj05h+ldAAAAAAAAAAAAAAAAAAAAAAAAAAAAFDs0aGW2IFefWp2YBTnMF3frtuDqpQECAyYgASFYIEdsXfmoFVtAxie5WbvcSZBbL/XIvpzRFx7VdoY3S1Q4IlggB/+Nb7vAg4sKmDF79GMTdTmxn0CFhZxpKV74WuTuFbc=';

  if (!isoBase64URL.isBase64url(attestationObject)) {
    console.log('converting base64 attestationObject to base64url');

    const atteObjBytes = isoBase64URL.toBuffer(attestationObject, 'base64');
    attestationObject = isoBase64URL.fromBuffer(atteObjBytes);
  }

  console.log('\nParsing native attestation object as base64url')
  const decoded1 = decodeAttestationObject(isoBase64URL.toBuffer(attestationObject));
  const parsedAuthData1 = parseAuthenticatorData(decoded1.get('authData'));
  console.log('response hex RP ID hash:', isoUint8Array.toHex(parsedAuthData1.rpIdHash));
  console.log(
    'computed hex RP ID hash:',
    isoUint8Array.toHex(
      await isoCrypto.digest(isoUint8Array.fromUTF8String('stageapi.getsecretstash.com'), COSEALG.ES256),
    ),
  );

  try {
    await matchExpectedRPID(parsedAuthData1.rpIdHash, ['stageapi.getsecretstash.com']);
    console.log('found match');
  } catch (err) {
    console.error(err);
  }
})();
```

And the output:

```
converting base64 attestationObject to base64url
Parsing native attestation object as base64url
response hex RP ID hash: 27cebafa48d4c6145a333be1938e3f4aca8e46de4d11f270f70a50ae3d3987e9
computed hex RP ID hash: 27cebafa48d4c6145a333be1938e3f4aca8e46de4d11f270f70a50ae3d3987e9
found match
```

**Reply by akozlik** [reply]
Awesome, I'm going to give this a shot. Now to just figure out how to access that isoBase64URL function outside of the library. I'll keep you up to date!

**Reply by akozlik** [reply]
LET'S GOOOOOOOOOO!!!!!!!

![Screenshot 2023-07-14 at 5 23 33 PM](https://github.com/MasterKale/SimpleWebAuthn/assets/467951/ebd22e29-432f-4bf8-87eb-762559ab68dc)

Thank you so much for the advice! I needed to use the HexagonBase64 library directly to do the translation from base64 to base64url

```
let attestationObject = body.attResp.response.attestationObject;
    var testObject = attestationObject.replace(/=/g, '');

    if (!HexagonBase64.validate(testObject, true)) {
        console.log('converting base64 attestationObject to base64url');
    
        const _buffer = HexagonBase64.toArrayBuffer(attestationObject, false);
        const atteObjBytes = new Uint8Array(_buffer);
        attestationObject = HexagonBase64.fromArrayBuffer(atteObjBytes, true);
        body.attResp.response.attestationObject = attestationObject;
      } else {
        console.log("Valid base 64");
      }
```

Passing that through to the validator worked!

Big lesson here that Apple uses Base64 natively but uses Base64URL through Safari. Converting from Base64 to Base64URL is exactly what was needed.

Matt I cannot tell you enough how much I appreciate your help this week. Please let me know if I can send some money your way through Venmo or something. Thank you so much! Heading into the weekend in a great mood!

**Reply by MasterKale** [reply]
> Now to just figure out how to access that isoBase64URL function outside of the library. 

There are a lot of "helpers", like `isoBase64URL` (and `isoUint8Array`, `isoCrypto`, `decodeAttestationObject()`, `parseAuthenticatorData()`, and many more) that are available for you to use for things like this. The trick is that you have to import them from `@simplewebauthn/server/helpers` instead:

```ts
import { isoBase64URL, isoUint8Array } from '@simplewebauthn/server/helpers';
```

Check those out, you can probably replace some dependencies with them.

> Matt I cannot tell you enough how much I appreciate your help this week. Please let me know if I can send some money your way through Venmo or something. Thank you so much! Heading into the weekend in a great mood!

Happy to help! And I'm always grateful for any support people like you throw my way. Here are the two ways I've got set up for people to financially support me and my work:

- **Github:** https://github.com/sponsors/MasterKale
- **Buy Me a Coffee:** https://www.buymeacoffee.com/iamkale

Both support monthly or one-time contributions for whatever level of support you deem appropriate 🙇

**Comment by spendres** [comment]
Try passing in "com.domain.stageapi" to the iOS library. I ran into a similar problem where their documentation said one form in one place and reverse domain in another. The iOS library may simply be verifying your plist entry and complaining that "com.domain.stageapi" != to "stageapi.domain.com";

**Reply by akozlik** [reply]
Interesting idea. We'll try it out but I'm not too confident. The odd thing is that the bytes look correct when we decide everything. It's a mystery for sure.

**Reply by akozlik** [reply]
No luck on this one. The reverse domain throws an error so it doesn't seem to be the culprit.

**Reply by LondonAtlas** [reply]
Thank you for the idea but this didn't work. Apple errored out saying 
`The operation couldn’t be completed. Application with identifier teamId.com.domain.appName is not associated with domain com.domain.stageapi`

**Comment by spendres** [comment]
Have you tried corbado's flutter iOS/app. At this point you need another native iOS library to compare. See:
https://github.com/corbado/example-passkeys-api-flutter

I was able to use the demo with a domain configured on apple developer, so it would help to eliminate your library – unless you are currently using it.

**Comment by DePasqualeOrg** [comment]
I'm trying to use this library with my native iOS app, and I'm passing the following data to the server for registration:

```typescript
type RequestData = {
  email?: string, // Base-64 encoded, the value the user enters in the email/user ID field
  credentialID: string, // Base-64 encoded
  clientDataJSON: string, // Base-64-encoded JSON, challenge also base-64 encoded: {"type":"webauthn.create","challenge":"cGFzc2tleSBjaGFsbGVuZ2U","origin":"https://<RPID>"}
  attestationObject: string, // Base-64 encoded
};
```

After looking at the examples in the documentation and trying the provided methods, I can't figure out how to use the server library to validate the registration. Could you provide an example?

**Comment by DePasqualeOrg** [comment]
Also it would be helpful to see an example of authentication with this data sent from an iOS app:

```typescript
type RequestData = {
  email?: string, // undefined
  userID: string, // Base-64 encoded, the value the user enters in the email/user ID field
  credentialID: string, // Base-64 encoded
  authenticatorData?: string, // Base-64 encoded
  clientDataJSON: string, // Base-64 encoded JSON, challenge base-64 encoded: {"type":"webauthn.get","challenge":"cGFzc2tleSBjaGFsbGVuZ2U","origin":"https://<RPID>"}
  signature?: string // Base-64 encoded
};
```
